[V1] Refactor parallel sampling support

markmc opened the pull request:
The initial implementation in #10980 went to great efforts to add parallel sampling as a wrapper at the highest layer of abstraction possible. This resulted in a lot of tricky code to post-process `RequestOutputs` to aggregate them where necessary.

Instead, it probably makes sense to implement parallel sampling at the layer that actually creates `RequestOutput` objects - i.e. in `OutputProcessor`

To do this, we simply need to allow for fanning out child requests in `LLMEngine.add_request()`, passing details of the fan-out to `OutputProcessor.`

This adds some overhead to the `n=1` case (see `SingularSamplingRequest`) in return for significantly less overhead and complication in the parallel sampling case.

markmc committed:
[V1][Parallel Sampling] Refactor recently added support

The initial implementation went to great efforts to add parallel
sampling as a wrapper at the highest layer of abstraction possible.
This resulted in a lot of tricky code to post-process RequestOutputs
to aggregate them where necessary.

Instead, it probably makes sense to implement parallel sampling at
the layer that actually creates RequestOutput objects - i.e. in
OutputProcessor.

To do this, we simply need to allow for fanning out child requests
in LLMEngine.add_request(), passing details of the fan-out to
OutputProcessor.

This adds some overhead to the n=1 case in return for significantly
less overhead and complication in the parallel sampling case.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -74,6 +85,94 @@ def from_new_request(
+        delta = self.output_kind == RequestOutputKind.DELTA
+
+        text = self.detokenizer.get_next_output_text(finished, delta)
+        if not delta:
+            token_ids = self.detokenizer.output_token_ids
+
+        logprobs = self.logprobs_processor.logprobs
+        if delta and logprobs:
+            logprobs = logprobs[-len(token_ids):]
```
suggested small simplification:
```suggestion
        logprobs = self.logprobs_processor.logprobs
        if not delta:
            token_ids = self.detokenizer.output_token_ids
        elif logprobs:
            logprobs = logprobs[-len(token_ids):]
```

markmc replied:
Yeah, this layout was deliberate though - first prepare the text/token_ids param using the detokenizer, second prepare the logprobs. Both are distinct steps that both happen to depend on delta mode

You could even imagine like:

```
  # Prepare text and token_ids, based on delta mode
  text, token_ids = self.detokenizer.get_next_output(finished, delta)

  # Prepare logprobs, based on delta mode
  logprobs = self.logprobs_processor.get_logprobs(delta, token_ids)
```

Would comments like that help (even without the new methods)?

Anyway ... not a big deal, but there was a reason for doing it this way

njhill replied:
Makes sense, was only minor suggestion but I follow your reasoning.

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -74,6 +85,94 @@ def from_new_request(
+        finished = finish_reason is not None
+        delta = self.output_kind == RequestOutputKind.DELTA
+
+        text = self.detokenizer.get_next_output_text(finished, delta)
```
Could move this down to the `CompletionOuput` constructor? (no need for intermediate var)

njhill replied:
Sure, again only a minor suggestion, I have a tendency towards making things more concise but that is usually quite subjective.

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -74,6 +85,94 @@ def from_new_request(
+        finished: bool,
+    ) -> RequestOutput:
+
+        delta = self.output_kind == RequestOutputKind.DELTA
+
+        if delta:
```
small simplification?
```suggestion
        if self.output_kind == RequestOutputKind.DELTA:
```

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -74,6 +85,94 @@ def from_new_request(
+        # In follow up, we will switch to invariant where EngineCore
+        # does not stream partial prefills.
+        if not finished and (self.is_prefilling
+                             or output_kind == RequestOutputKind.FINAL_ONLY):
```
Have an `final_only` var that can be used here and below?

markmc replied:
Sounds good :+1:

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -74,6 +85,94 @@ def from_new_request(
+            new_token_ids, finish_reason, stop_reason)
+        request_output.outputs.append(completion_output)
+
+        if (aggregate_children
+                and len(request_output.outputs) != self.parent_req.n):
```
Would just need to add a `self.parent_req is not None` condition here I think, if adopting the suggestion about that above.

markmc replied:
Yep :+1: 

But also `self.parent_req.request_id` above

markmc replied:
Done now :+1:

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -45,6 +50,8 @@ def __init__(
         self.is_prefilling = True
         self.queue = queue
 
+        self.output_aggregator: Optional[RequestOutput] = None
```
Add comment here?

markmc replied:
Sure, will do :+1: 

```
	# Used to aggregate child completions when not streaming                                                                                                         
        self.output_aggregator: Optional[RequestOutput] = None
````

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -144,25 +144,26 @@ async def add_request(
         """Add new request to the AsyncLLM."""
 
         # 1) Create a new output queue for the request.
-        if self.output_processor.is_request_active(request_id):
-            raise ValueError(f"Request id {request_id} already running.")
```
Curious of the reason for removing this check?

markmc replied:
I think I was pondering whether `is_request_active()` should check the parent request ID or the child request ID, and then I noticed in `OutputProcessor`:

```
    def add_request(
        self,
	request: EngineCoreRequest,
        parent_req: ParentRequest,
        request_index: int,
	queue: Optional[asyncio.Queue[RequestOutput]] = None,
    ) -> None:
	request_id = request.request_id
        if request_id in self.request_states:
            raise ValueError(f"Request id {request_id} already running.")
```

So it seems completely redundant?

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -144,25 +144,26 @@ async def add_request(
-                                                prompt_adapter_request,
-                                                priority)
+        # 2) Fan out child requests (for n>1)
+        parent_req = ParentRequest.from_params(request_id, params)
+        for idx in range(parent_req.n):
+            child_req_id, child_params = parent_req.get_child_info(idx)
```
WDTY about adjusting this and making `RequestState.parent_req` `Optional`, so that we won't need `SingularSamplingRequest`? Mainly to reduce the abstractions and make the logic here a bit more transparent.
```suggestion
        parent_req = ParentRequest.from_params(request_id, params) if params.n > 1 else None
        for idx in range(params.n):
            if parent_req:
                request_id, params = parent_req.get_child_info(idx)
```

markmc replied:
All great points Nick, thanks for pushing on this - I think the end result is better for it :+1:

njhill reviewed (commented):
Thanks @markmc, looks great and I agree doing the aggregation in the output processor is much nicer! Just some minor suggestions to simplify further.

markmc commented on vllm/v1/engine/output_processor.py:
```diff
@@ -79,6 +91,93 @@ def from_new_request(
+        else:
+            prompt_logprobs = self.logprobs_processor.prompt_logprobs
+
+        return RequestOutput(
```
Right, it's set to None by default, and we're not currently creating a `RequestMetrics` object in the v1 code path :+1:

afeldman-nm replied:
Yes sorry - upon further study of how metrics work, I realize that this field is actually not relevant :)

markmc commented on vllm/v1/engine/output_processor.py:
```diff
@@ -79,6 +91,93 @@ def from_new_request(
+        request_output.outputs.append(completion_output)
+
+        if (final_only and len(request_output.outputs) != self.parent_req.n):
+            self.output_aggregator = request_output
```
Nope, that was my silly refactoring mistake (one quick, last, "safe" refactor after testing!) - the output aggregator needs to be on the ParentRequest not the RequestState ... since the point is to aggregate across child requests!

afeldman-nm commented:
Initial benchmarking results with `n=1` (averaged over 3x runs) - it appears that there is not a performance regression. I may try a serving benchmark later.

`main`:

* Throughput: 41.7 rps
* P99 `LLMEngine` average execution time: 2.50 sec

`parallel-sampling-refactor`:

* Throughput (rps): 42.0 rps
* P99 `LLMEngine` average execution time: 2.47 sec

Command to measure `LLMEngine` throughput:

```
VLLM_USE_V1=1 VLLM_ENABLE_V1_MULTIPROCESSING=1 python benchmarks/benchmark_throughput.py --model meta-llama/Llama-3.2-3B-Instruct --input-len 128 --output-len 512 --trust-remote-code
```

Command to measure `LLMEngine` average execution time:

```
VLLM_USE_V1=1 VLLM_ENABLE_V1_MULTIPROCESSING=1 python benchmarks/benchmark_latency.py --model meta-llama/Llama-3.2-3B-Instruct --input-len 128 --output-len 512 --trust-remote-code
```

afeldman-nm commented:
Async serving benchmarking results with `n=1` (averaged over 3x runs) - it appears that there is not a performance regression. Since serving benchmark results can have high variance on beaker, for each metric I also include the different between the "worst" and "best" value of each metric (as a percentage of the "best").

`main`:

* Throughput: 79.44 rps (Lowest-Highest %: -2.3%)
* P99 TTFT: 3512.56 ms (Highest-Lowest %: 5.3%)
* P99 TPOT: 110.70ms (Highest-Lowest %: 21.6%)
* P99 ITL: 114.84ms (Highest-Lowest %: 3.0%)

`parallel-sampling-refactor`:

* Throughput: 80.49 rps (Lowest-Highest %: -5.3%), +1% improvement over `main`
* P99 TTFT: 3289.69 ms (Highest-Lowest %: 6.4%), -6% improvement over `main`
* P99 TPOT: 98.53ms (Highest-Lowest %: 1.52%), -11% improvement over `main`
* P99 ITL: 111.13ms (Highest-Lowest %: 0.96%), -3% improvement over `main`

Command to bring up vLLM v1 engine server:

```
VLLM_USE_V1=1 VLLM_ENABLE_V1_MULTIPROCESSING=1 vllm serve meta-llama/Llama-3.2-3B-Instruct --port 8091 --disable-log-requests --no-enable-prefix-caching
```

Benchmark command:

```
python benchmarks/benchmark_serving.py --model meta-llama/Llama-3.2-3B-Instruct --dataset-path ../sharegpt.json  --port 8091
```

afeldman-nm commented:
> Nice work cleaning this up!
> 
> > This adds some overhead to the n=1 case (see SingularSamplingRequest) in return for significantly less overhead and complication in the parallel sampling case.
> 
> We should verify that this overhead is negligible with a quick benchmark

Based on the benchmarking results above, there does not appear to be a perf regression.

afeldman-nm commented on vllm/v1/engine/parallel_sampling.py:
```diff
@@ -1,69 +1,46 @@
 
     request_id: str
-    sampling_params: SamplingParams
+    params: Union[SamplingParams, PoolingParams]
```
I don't think `PoolingParams` supports `n`. Do we need to support `PoolingParams` in `ParentRequest`?

markmc replied:
Yep, good point. I've made `from_params()` continue to support both, but made the constructor only support `SamplingParams`. Nice cleanup, thanks :+1:

afeldman-nm commented on vllm/v1/engine/parallel_sampling.py:
```diff
@@ -1,69 +1,46 @@
+    output_aggregator: Optional[RequestOutput]
+
+    # To efficiently obtain child sampling params
     cached_child_sampling_params: Optional[SamplingParams]
```
Two thoughts here:
1. if we are changing `sampling_params` to `params`, should we change this member to `cached_child_params`?
2. See my other comment about `PoolingParams` - if `params` can be `PoolingParams`, wouldn't `cached_child_params` need to accommodate `PoolingParams`?

afeldman-nm commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -156,48 +152,25 @@ def add_request(
-        self.engine_core.add_request(request)
+        # 1) Fan out child requests (for n>1)
+        parent_req = ParentRequest.from_params(request_id, params)
+        n = params.n if isinstance(params, SamplingParams) else None
+        for idx in range(n):
```
I agree that here we need to accommodate the possibility that `params` is not a `SamplingParams` instance - i.e. if `params` is a `PoolingParams` instance - however, if `n` is assigned `None` then won't that break `range(n)` in `for idx in range(n)`?

Perhaps `n = params.n if isinstance(params, SamplingParams) else 1` ?

njhill replied:
Yes it should be `1` here
```suggestion
        n = params.n if isinstance(params, SamplingParams) else 1
        for idx in range(n):
```

markmc replied:
Great catch! It was correct in `async_llm.py`. Fixed now :+1:

afeldman-nm commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -156,48 +152,25 @@ def add_request(
+                request_id, params = parent_req.get_child_info(idx)
+
+            # 2) Process raw inputs into the request.
+            request = self.processor.process_inputs(request_id, prompt, params,
+                                                    arrival_time, lora_request,
+                                                    trace_headers,
+                                                    prompt_adapter_request,
+                                                    priority)
```
Nit: kwargs? i.e.

```
request = self.processor.process_inputs(
  request_id,
  prompt=prompt,
  params,
  arrival_time=arrival_time,
  lora_request=lora_request,
  trace_headers=trace_headers,
  prompt_adapter_request=prompt_adapter_request,
  priority=priority)
```

afeldman-nm commented on vllm/v1/engine/async_llm.py:
```diff
@@ -144,25 +144,30 @@ async def add_request(
-        # 3) Add the request to OutputProcessor (this process).
-        self.output_processor.add_request(request, queue)
+            # 3) Convert Input --> Request.
+            request = self.processor.process_inputs(request_id, prompt, params,
+                                                    arrival_time, lora_request,
+                                                    trace_headers,
+                                                    prompt_adapter_request,
+                                                    priority)
```
Something I wondered - can we reuse the output of `process_inputs` over multiple child requests? Since I believe `process_inputs` operates solely on the prompt (or MM inputs or whatever.)

This could be tricky for multimodal so I don't know if it is in-scope for this PR.

markmc replied:
Yeah, I did wonder that too, but didn't dig into it. Seem quite orthogonal to this PR though

njhill replied:
Yes agree. Actually we can move this outside of the loop I think, and just update the request_id and sampling_params of the request inside the loop.

Doesn't need to hold up this PR but would be quite a simple change I think.

afeldman-nm commented on vllm/v1/engine/output_processor.py:
```diff
@@ -79,6 +88,90 @@ def from_new_request(
+        if self.parent_req is None:
+            return append_completion_output(new_request_output(
+                self.request_id))
+        else:
+            return self.parent_req.make_request_output(
+                final_only, new_request_output, append_completion_output)
```
Nit: redundant else (since the if-body returns)

markmc replied:
Thanks. It was a stylistic choice, but I've made the change :+1:

afeldman-nm commented on vllm/v1/engine/parallel_sampling.py:
```diff
@@ -149,227 +86,32 @@ def get_child_info(self, index: int) -> Tuple[str, SamplingParams]:
+        new_request_output: Callable[[str], RequestOutput],
+        append_completion_output: Callable[[RequestOutput], RequestOutput],
+    ) -> Optional[RequestOutput]:
+        # Use an existing RequestOutput if we're aggregating
+        request_output = self.output_aggregator
 
+        # Make new RequestOutput otherwise
+        if request_output is None:
+            request_output = new_request_output(self.request_id)
```
I *think* you could refactor this slightly as so:

```
if not (request_output := self.output_aggregator):
    request_output = new_request_output(self.request_id)
```

markmc replied:
This aggregator logic is a little tricky to follow, so I think the separate line and comment helps make the flow understandable

afeldman-nm commented on vllm/v1/engine/output_processor.py:
```diff
@@ -79,6 +88,90 @@ def from_new_request(
+            token_ids = self.detokenizer.output_token_ids
+
+        # Prepare logprobs, based on delta mode
+        logprobs = self.logprobs_processor.logprobs
+        if delta and logprobs:
+            logprobs = logprobs[-len(token_ids):]
```
Maybe not worth it, but I think you could replace this with

if delta and (logprobs := self.logprobs_processor.logprobs):
    logprobs = logprobs[-len(token_ids):]

markmc replied:
Yeah, I think that would hurt readability, even if it's more condensed

afeldman-nm reviewed (changes requested):
Thanks for the PR @markmc ! I left some feedback.

njhill commented on vllm/v1/engine/parallel_sampling.py:
```diff
@@ -149,227 +86,32 @@ def get_child_info(self, index: int) -> Tuple[str, SamplingParams]:
-        gens.append(gen)
+        # If not streaming, aggregate until all child requests complete
+        if (final_only and len(request_output.outputs) != self.n):
+            self.output_aggregator = request_output
```
It would make more sense to me for this assignment to be in the other if above.

njhill replied:
Ah right, ok ignore my comment then! I'm not sure why I find it slightly confusing to overwrite this for the intermediate outputs.

njhill commented on vllm/v1/engine/parallel_sampling.py:
```diff
@@ -149,227 +86,32 @@ def get_child_info(self, index: int) -> Tuple[str, SamplingParams]:
-        gen = parent_req.wrap_child_async_generator(child_gen, idx)
-        gens.append(gen)
+        # If not streaming, aggregate until all child requests complete
+        if (final_only and len(request_output.outputs) != self.n):
```
nit: redundant parentheses

markmc commented on vllm/v1/engine/llm_engine.py:
```diff
@@ -156,48 +152,25 @@ def add_request(
+                request_id, params = parent_req.get_child_info(idx)
+
+            # 2) Process raw inputs into the request.
+            request = self.processor.process_inputs(request_id, prompt, params,
+                                                    arrival_time, lora_request,
+                                                    trace_headers,
+                                                    prompt_adapter_request,
+                                                    priority)
```
This PR just indents this existing code, so I think that would be an orthogonal change ... and not even sure I prefer it (given we have type checking)

markmc commented on vllm/v1/engine/async_llm.py:
```diff
@@ -144,25 +144,30 @@ async def add_request(
-        # 3) Add the request to OutputProcessor (this process).
-        self.output_processor.add_request(request, queue)
+            # 3) Convert Input --> Request.
+            request = self.processor.process_inputs(request_id, prompt, params,
+                                                    arrival_time, lora_request,
+                                                    trace_headers,
+                                                    prompt_adapter_request,
+                                                    priority)
```
Same as above, orthogonal, not sure I prefer it

markmc commented on vllm/v1/engine/parallel_sampling.py:
```diff
@@ -1,69 +1,46 @@
+    def from_params(
+        cls,
+        request_id: str,
+        params: Union[SamplingParams, PoolingParams],
```
Supporting it here is useful yes, but good idea to remove it elsewhere

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -79,6 +88,89 @@ def from_new_request(
+        def new_request_output(request_id: str) -> RequestOutput:
+            return self._new_request_output(request_id, finished)
+
+        def append_completion_output(
+                request_output: RequestOutput) -> RequestOutput:
+            request_output.outputs.append(
+                self._new_completion_output(new_token_ids, finish_reason,
+                                            stop_reason))
+            return request_output
```
I'm not sure we need this method, could we just create the completion output here and pass it directly to `parent_req.make_request_output`?

markmc replied:
Ok, done :+1:

njhill reviewed (commented):
Thanks @markmc looks great, just a couple more small things.

njhill reviewed (approved):
LGTM

markmc committed:
[V1][Parallel Sampling] Fix test_output_processor.py

Add default values for OutputProcessor.add_request() parent_req
and index parameters to fix tests.

These are the values used in the n==1 case.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
Lint and deploy minio setup failing with:

```
make_bucket failed: s3://testbucket Could not connect to the endpoint URL: "http://127.0.0.1:9000/testbucket"
```

markmc committed:
[V1][Parallel Sampling] Sort completion outputs by index

As per the original V1 parallel sampling code.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>
